### PR TITLE
Add terraform onboarding for prometheus addon

### PR DESCRIPTION
**Pull request by Sohamdg081992**
Deployed the resources in test resource group : https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/ce4d1293-71c0-4c72-bc55-133553ee9e50/resourceGroups/rg-glorious-redbird/overview

### CODE REVIEW

File: AddonTerraformTemplate/main.tf [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        ]
+      }
+  })
+
+  schema_validation_enabled = false
+  ignore_missing_property   = false
+}
+
+resource "azapi_resource" "KubernetesReccordingRulesRuleGroup" {
+  type = "Microsoft.AlertsManagement/prometheusRuleGroups@2021-07-22-preview"
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:173]: Update to use GA api. See PR : https://github.com/Azure/prometheus-collector/pull/446

File: AddonTerraformTemplate/main.tf [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        ]
+      }
+  })
+
+  schema_validation_enabled = false
+  ignore_missing_property   = false
+}
+
+resource "azapi_resource" "NodeRecordingRulesRuleGroupWin" {
+  type = "Microsoft.AlertsManagement/prometheusRuleGroups@2021-07-22-preview"
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:282]: GA Api : https://github.com/Azure/prometheus-collector/pull/446

File: AddonTerraformTemplate/main.tf [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        ]
+      }
+  })
+  schema_validation_enabled = false
+  ignore_missing_property   = false
+}
+
+
+resource "azapi_resource" "NodeAndKubernetesRecordingRulesRuleGroupWin" {
+  type = "Microsoft.AlertsManagement/prometheusRuleGroups@2021-07-22-preview"
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:363]: GA Api : https://github.com/Azure/prometheus-collector/pull/446

File: AddonTerraformTemplate/main.tf [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+resource "azurerm_dashboard_grafana" "example" {
+  name                              = "example-dg"
+  resource_group_name               = azurerm_resource_group.rg.name
+  location                          = azurerm_resource_group.rg.location
+
+  identity {
+    type = "SystemAssigned"
+  }
+
+  azure_monitor_workspace_integrations {
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:101]: Does this overwrite existing azure monitor workspace integrations for the grafana workspace or append to it?

File: AddonTerraformTemplate/main.tf [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+    resource_id  = var.monitor_workspace_id
+  }
+
+}
+
+resource "azapi_resource" "NodeRecordingRulesRuleGroup" {
+  type = "Microsoft.AlertsManagement/prometheusRuleGroups@2023-03-01"
+  name = "example-NodeRecordingRulesRuleGroup"
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:109]: Any reason as to why the name's starting with 'example-'? We should keep the default names consistant with the other forms of onboarding (ARM, Bicep, Policy, CLI, UX) i.e. 'NodeRecordingRulesRuleGroup' imo

File: AddonTerraformTemplate/main.tf [hunk 6]
```diff
@@ -0,0 +1,448 @@
+# Generate random resource group name
+resource "random_pet" "rg_name" {
```
Comment (bragi92) [inline AddonTerraformTemplate/main.tf:2]: What is this used for?

File: AddonTerraformTemplate/main.tf [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+
+  default_node_pool {
+    name       = "agentpool"
+    vm_size    = "Standard_D2_v2"
+    node_count = var.agent_count
+  }
+
+  monitor_metrics {
+    annotations_allowed = "string"
+    labels_allowed = "string"
```
Comment (gracewehner) [inline AddonTerraformTemplate/main.tf:30]: Thanks Soham, could we make the annotations_allowed and labels_allowed as parameters with the default value as an empty string, to match the ARM and CLI onboarding? Also do you have a link to the terraform documentation for this? Who adds the terraform support for the new azure monitor metrics profile?

File: AddonTerraformTemplate/main.tf (hunk 1 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf:173]: Updated. Thanks Kaveesh!

File: AddonTerraformTemplate/main.tf (hunk 4 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf:101]: yes, this overrides the existing azure monitor workspace integrations, similar to other templates(policy,arm,etc.)

File: AddonTerraformTemplate/main.tf (hunk 6 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf:2]: removed

File: AddonTerraformTemplate/main.tf (hunk 5 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf:109]: Thanks Kaveesh! Renaming this to 'NodeRecordingRulesRuleGroup' as discussed.

File: AddonTerraformTemplate/main.tf (hunk 7 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf]: Thanks Grace! I parameterized the annotations_allowed and labels_allowed similar to the other templates, in the latest revision. Here is the [link](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#monitor_metrics) for the docs. These resources are added in terraform by the swagger files. The respective teams are responsible for keeping the swagger files updated in Terraform, it is AKS team for the new azure monitor metrics profile.

**Review by gracewehner — APPROVED** [review]

**Review by bragi92 — APPROVED** [review]

**Comment by gracewehner** [comment]
Thanks Soham, also when you merge could you choose "squash and merge"? This just makes the commit history in main easier to see which PRs go in and when :)
image

**Pull request merged**